drivers/include/drivers/interfaces/InterfaceDigitalInOut.h

```diff
@@ -0,0 +1,72 @@
+/*
+ * Mbed-OS Microcontroller Library
+ * Copyright (c) 2021 Embedded Planet
+ * Copyright (c) 2021 ARM Limited
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License
+ */
+
+#ifndef MBED_INTERFACE_DIGITALINOUT_H_
+#define MBED_INTERFACE_DIGITALINOUT_H_
+
+#include "PinNames.h"
+
+namespace mbed {
+
+class DigitalInOut;
+
+namespace interface {
+
+#ifdef FEATURE_EXPERIMENTAL_API
+
+// TODO - move method doxygen comments to interface once this polymorphism is mainstream
+
+// Pure interface definition for DigitalInOut
+struct DigitalInOut {
+    virtual ~DigitalInOut() = default;
+    virtual void write(int value) = 0;
+    virtual int read() = 0;
+    virtual void output() = 0;
+    virtual void input() = 0;
+    virtual void mode(PinMode pull) = 0;
+    virtual int is_connected() = 0;
+
+    virtual DigitalInOut &operator= (int value)
```

Comment:
> a reference to an interface can't be returned, right?

Sure it can. That's the polymorphism. 

As virtual functions, these would be the default implementations in the vtable that the subclasses would get, so they would be called via any interface or concrete reference in the absence of an override.

But as both @pan- and I noted, the virtual doesn't really make sense here, as you don't want subclasses to override. You always want this definition.

Given 

```
     interface::DigitalOut &x;

     x = 1;
```

that could statically use `interface::DigitalOut::operator=` which will then look up the vtable to call the virtual `write` for whatever the real class is.

Having `operator=` not be virtual saves a vtable lookup+entry for the `operator=`, but it does impose the requirement that the class should not attempt to override - `DigitalOut::operator=` should be the inherited `interface::DigitalOut::operator=` so `=` works the same for interface or concrete reference (call `write`).

In the current PR, `operator=` is given the unwanted leeway to do something other than call `write` in derived classes.

I don't think you can stop someone overriding a non-virtual method, but most compilers will warn if they see you overriding a non-virtual method in a virtual class. And if you tried using the `override` keyword on the override, that would block you - so the good style of always including `override` is a safety check, although that's not done here (you'd need an MBED_OVERRIDE_IF_EXPERIMENTAL macro or whatever to do that).

Although - what if this was a `final` function? Normally `final` is used to "seal" an actually-virtual vtable method. What do compilers do if a method is just `final` in the base class? Do they avoid the vtable entry totally? If they do, that's what you want. "This is virtual+final in a language sense, so you can't override it, but it's totally implemented like a non-virtual function".